chipsec/hal/interrupts.py

```diff
@@ -140,3 +140,53 @@ def send_ACPI_SMI(self, thread_id, smi_num, buf_addr, invoc_reg, guid, data):
             #clear CommBuffer
             self.cs.helper.write_physical_mem(buf_addr,len(data_hdr),"\x00"*len(data_hdr))
             return None
+
+    # scan phys mem range start-end looking for 'smmc'
+    def scan_range(self,start,end):
+        chunk_sz = 1024 * 8 #8KB chunks
+        phys_address = start
+        found_at = 0
+        while phys_address <= end:
+            buffer = self.cs.mem.read_physical_mem( phys_address, chunk_sz )
+            buffer = chipsec.defines.bytestostring(buffer)
+            offset = buffer.find('smmc')
+            if offset != -1:
+                found_at = phys_address + offset
+                break
+            phys_address += chunk_sz
+        return found_at
+
+    
+    '''
+Send SWSMI in the same way as EFI_SMM_COMMUNICATION_PROTOCOL
+    - Write Commbuffer location and Commbuffer size to 'smmc' structure
+    - Write 0 to 0xb3 and 0xb2
+
+MdeModulePkg/Core/PiSmmCore/PiSmmCorePrivateData.h
+
+#define SMM_CORE_PRIVATE_DATA_SIGNATURE  SIGNATURE_32 ('s', 'm', 'm', 'c')
+ struct {
+  UINTN                           Signature;
+   This field is used by the SMM Communicatioon Protocol to pass a buffer into
+   a software SMI handler and for the software SMI handler to pass a buffer back to
+   the caller of the SMM Communication Protocol.
+  VOID                            *CommunicationBuffer;
+  UINTN                           BufferSize;
+
+  EFI_STATUS                      ReturnStatus;
+} SMM_CORE_PRIVATE_DATA;
+    '''
+    def send_smmc_SMI(self,gSmmCorePrivate,guid,payload,payload_loc):
+        guid_b = uuid.UUID(guid).bytes_le
+        payload_sz = len(payload)
+
+
+        data_hdr = guid_b + struct.pack("Q",payload_sz) + payload
+        # write payload to payload_loc
+        CommBuffer_offset = 56
+        BufferSize_offset = CommBuffer_offset + 8
+        self.cs.mem.write_physical_mem(gSmmCorePrivate + CommBuffer_offset,8,struct.pack("Q",payload_loc))
+        self.cs.mem.write_physical_mem(gSmmCorePrivate + BufferSize_offset,8,struct.pack("Q",payload_sz))
+        self.cs.mem.write_physical_mem(payload_loc,len(data_hdr),data_hdr)
+        self.send_SMI_APMC(0x0,0x0)
+
```

Comment:
Will a return value be put back in the buffer and is it worthwhile to read?

Reply:
Yeah, there will be a return code... I should return that. Will add